utils/background-task-queue.ts

```diff
@@ -0,0 +1,232 @@
+ * // a2 will start only after a1 finishes, even when we don't await a1.run()
+ * a2.run().then((result) => console.assert(result === 15));
+ * ```
+ */
+export class BackgroundTaskQueue<M extends TaskModule> {
+  private modulePath: string;
+  private name: string;
+
+  private worker: Worker;
+  private lock = new Lock();
+  private isActivated = false;
+  private jobCounter = 0;
+
+  constructor(jobDescriptionModulePath: string, queueName: string) {
+    this.modulePath = jobDescriptionModulePath;
+    this.name = queueName;
+    const __filename = legacyFilename(import.meta);
+    this.worker = new Worker(__filename, { stderr: true, stdout: true });
+  }
+
+  add(input: Parameters<M["default"]>[0] = {}) {
+    type ThenArg<T> = T extends PromiseLike<infer U> ? U : T;
+    type RetType = ThenArg<ReturnType<M["default"]>>;
+
+    const id = this.generateId();
+
+    const log = new Logger(id, input);
+    log.markTime("init");
+    let hasRun = false;
+
+    const run = async () => {
+      log.markTime("queue");
+      if (hasRun) {
+        throw new Error("Cannot run the same job again.");
+      }
+      hasRun = true;
+
+      if (!this.isActivated) {
+        await this.activate();
+      }
+
+      await this.lock.acquire();
+      log.markTime("start");
+
+      const msg: CallRequest = { id, type: "call", input };
+      this.worker.postMessage(msg);
+
+      this.worker.stdout.pipe(split2()).on("data", log.onstdout);
+      this.worker.stderr.pipe(split2()).on("data", log.onstderr);
+
+      try {
+        return await new Promise<RetType>((resolve, reject) => {
+          const listener = (response: Response) => {
+            if (response.id === id) {
+              this.lock.release();
+              this.worker.removeListener("message", listener);
+
+              const { type, result } = response;
+
+              log.setResult(type, result);
+              this.worker.stdout.off("data", log.onstdout);
+              this.worker.stderr.off("data", log.onstderr);
+              log.markTime("finish");
+
+              type === "success" ? resolve(result as RetType) : reject(result);
+            }
+          };
+          this.worker.addListener("message", listener);
+        });
+      } finally {
+        await log.write();
+      }
+    };
+
+    return { id, run };
+  }
+
+  private async activate() {
+    this.isActivated = true;
+
+    const id = nanoid();
+    const modulePath = this.modulePath;
+    const msg: OperationRequest = { id, type: "init", modulePath };
+    this.worker.postMessage(msg);
+
+    await new Promise<void>((resolve, reject) => {
+      this.worker.once("message", (response: Response) => {
+        if (response.id !== id) {
+          reject(new Error(`Failed to register worker module: ${modulePath}`));
+        } else {
+          if (response.result === "success") {
+            resolve();
+          } else {
+            reject(response.result);
```

Comment:
is result always a real `Error` in this case?

Reply:
Can be any value the function throws.

Reply:
Ok, might be worth wrapping in an `Error`, just to get the stack trace.